lxd/apparmor/rsync.go

```diff
@@ -0,0 +1,192 @@
+package apparmor
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/pborman/uuid"
+
+	"github.com/lxc/lxd/lxd/revert"
+	"github.com/lxc/lxd/lxd/sys"
+	"github.com/lxc/lxd/shared"
+)
+
+var rsyncProfileTpl = template.Must(template.New("rsyncProfile").Parse(`#include <tunables/global>
+profile "{{ .name }}" flags=(attach_disconnected,mediate_deleted) {
+  #include <abstractions/base>
+
+  capability dac_override,
+  capability dac_read_search,
+  capability mknod,
+  capability chown,
+  capability fsetid,
+  capability fowner,
+
+  unix (connect, send, receive) type=stream,
+
+  @{PROC}/@{pid}/cmdline r,
+  @{PROC}/@{pid}/cpuset r,
+  {{ .rootPath }}/{etc,lib,usr/lib}/os-release r,
+
+  {{ .logPath }}/*/netcat.log rw,
+
+  {{ .rootPath }}/run/{resolvconf,NetworkManager,systemd/resolve,connman,netconfig}/resolv.conf r,
+  {{ .rootPath }}/run/systemd/resolve/stub-resolv.conf r,
+
+{{- if .sourcePath }}
+  {{ .sourcePath }}/** r,
+  {{ .sourcePath }}/ r,
+{{- end }}
+
+{{- if .dstPath }}
+  {{ .dstPath }}/** rw,
+  {{ .dstPath }}/ rw,
+{{- end }}
+
+{{- if .snap }}
+  # Snap-specific libraries
+  /snap/lxd/*/lib/**.so* mr,
+{{- end }}
+
+  {{ .execPath }} mixr,
+
+{{if .libraryPath -}}
+  # Entries from LD_LIBRARY_PATH
+{{range $index, $element := .libraryPath}}
+  {{$element}}/** mr,
+{{- end }}
+{{- end }}
+
+  deny {{ .rootPath }}/etc/ssl/openssl.cnf r,
+  deny /sys/devices/virtual/dmi/id/product_uuid r,
+  deny /sys/kernel/mm/transparent_hugepage/hpage_pmd_size r,
+}
+`))
+
+// RsyncWrapper is used as a RunWrapper in the rsync package.
+func RsyncWrapper(sysOS *sys.OS, cmd *exec.Cmd, sourcePath string, dstPath string) (func(), error) {
+	if !sysOS.AppArmorAvailable {
+		return func() {}, nil
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Attempt to deref all paths.
+	if sourcePath != "" {
+		fullPath, err := filepath.EvalSymlinks(sourcePath)
+		if err == nil {
+			sourcePath = fullPath
+		}
+	}
+
+	if dstPath != "" {
+		fullPath, err := filepath.EvalSymlinks(dstPath)
+		if err == nil {
+			dstPath = fullPath
+		}
+	}
+
+	// Load the profile.
+	profileName, err := rsyncProfileLoad(sysOS, sourcePath, dstPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load rsync profile: %w", err)
+	}
+
+	revert.Add(func() { _ = unloadProfile(sysOS, profileName, profileName) })
+	revert.Add(func() { _ = deleteProfile(sysOS, profileName, profileName) })
```

Comment:
deleteProfile calls unloadProfile internally so I dont think we need the call to unloadProfile here.